File: apps/macos/LuminaTests/Unit/ColdStartReadinessTests.swift
import XCTest
@testable import Lumina

final class ColdStartReadinessTests: XCTestCase {
    func testProductReadyRequiresEngineDataCacheOnly() {
        var snap = ColdStartPhaseSnapshot(
            engine: .done,
            data: .done,
            cache: .done,
            news: .running
        )
        XCTAssertTrue(ColdStartReadiness.isProductReady(snap))
        snap.cache = .running
        XCTAssertFalse(ColdStartReadiness.isProductReady(snap))
        snap.cache = .done
        snap.news = .failed
        XCTAssertTrue(ColdStartReadiness.isProductReady(snap))
        XCTAssertTrue(ColdStartReadiness.isBootNewsTerminal(.failed))
        XCTAssertFalse(ColdStartReadiness.isBootNewsTerminal(.running))
    }

    func testMergeMarksEngineRunningUntilDone() {
        let pending = ColdStartReadiness.merge(
            engineDone: false,
            data: "pending",
            cache: "pending",
            news: "pending",
            newsDetail: nil
        )
        XCTAssertEqual(pending.engine, .running)
        let ready = ColdStartReadiness.merge(
            engineDone: true,
            data: "done",
            cache: "running",
            news: "pending",
            newsDetail: nil
        )
        XCTAssertEqual(ready.engine, .done)
        XCTAssertEqual(ready.data, .done)
        XCTAssertEqual(ready.cache, .running)
    }

    func testDetailRevealPolicy() {
        XCTAssertEqual(ColdStartReadiness.detailRevealAfterSeconds, 10)
        XCTAssertFalse(ColdStartReadiness.shouldRevealTechnicalDetail(elapsedSeconds: 9.9))
        XCTAssertTrue(ColdStartReadiness.shouldRevealTechnicalDetail(elapsedSeconds: 10))
    }

    func testTechnicalDetailDerivesFromInternalPhase() {
        let engine = ColdStartReadiness.merge(
            engineDone: false,
            data: "pending",
            cache: "pending",
            news: "pending",
            newsDetail: nil
        )
        XCTAssertEqual(ColdStartReadiness.technicalDetail(engine), "正在启动引擎")

        let data = ColdStartReadiness.merge(
            engineDone: true,
            data: "running",
            cache: "pending",
            news: "pending",
            newsDetail: nil
        )
        XCTAssertEqual(ColdStartReadiness.technicalDetail(data), "正在准备阅读数据")

        let cache = ColdStartReadiness.merge(
            engineDone: true,
            data: "done",
            cache: "running",
            cacheProgress: 0.45,
            cacheDetail: "恢复书籍状态 (12/48)",
            news: "pending",
            newsDetail: nil
        )
        XCTAssertEqual(
            ColdStartReadiness.technicalDetail(cache),
            "恢复书籍状态 (12/48)"
        )

        let cachePlain = ColdStartReadiness.merge(
            engineDone: true,
            data: "done",
            cache: "running",
            news: "pending",
            newsDetail: nil
        )
        XCTAssertEqual(ColdStartReadiness.technicalDetail(cachePlain), "正在加载缓存")

        XCTAssertEqual(
            ColdStartReadiness.technicalDetail(cachePlain, launchError: "无法连接"),
            "无法连接"
        )
    }
}

File: apps/windows/Lumina.Tests/ColdStartReadinessTests.cs
using Lumina.Services;
using Xunit;

namespace Lumina.Tests;

public class ColdStartReadinessTests
{
    [Fact]
    public void IsProductReady_ignores_news()
    {
        var snap = ColdStartReadiness.Merge(true, "done", "done", "running", null);
        Assert.True(ColdStartReadiness.IsProductReady(snap));
        snap = ColdStartReadiness.Merge(true, "done", "running", "pending", null);
        Assert.False(ColdStartReadiness.IsProductReady(snap));
        snap = ColdStartReadiness.Merge(true, "done", "done", "failed", "timeout");
        Assert.True(ColdStartReadiness.IsProductReady(snap));
        Assert.True(ColdStartReadiness.IsBootNewsTerminal(ColdStartPhaseState.Failed));
        Assert.False(ColdStartReadiness.IsBootNewsTerminal(ColdStartPhaseState.Running));
    }

    [Fact]
    public void Detail_reveal_policy()
    {
        Assert.Equal(TimeSpan.FromSeconds(10), ColdStartReadiness.DetailRevealAfter);
        Assert.False(ColdStartReadiness.ShouldRevealTechnicalDetail(TimeSpan.FromSeconds(9.9)));
        Assert.True(ColdStartReadiness.ShouldRevealTechnicalDetail(TimeSpan.FromSeconds(10)));
    }

    [Fact]
    public void TechnicalDetail_derives_from_internal_phase()
    {
        var engine = ColdStartReadiness.Merge(false, "pending", "pending", "pending", null);
        Assert.Equal("正在启动引擎", ColdStartReadiness.TechnicalDetail(engine));

        var data = ColdStartReadiness.Merge(true, "running", "pending", "pending", null);
        Assert.Equal("正在准备阅读数据", ColdStartReadiness.TechnicalDetail(data));

        var cache = ColdStartReadiness.Merge(
            true, "done", "running", "pending", null, 0.45, "恢复书籍状态 (12/48)");
        Assert.Equal("恢复书籍状态 (12/48)", ColdStartReadiness.TechnicalDetail(cache));

        var cachePlain = ColdStartReadiness.Merge(true, "done", "running", "pending", null);
        Assert.Equal("正在加载缓存", ColdStartReadiness.TechnicalDetail(cachePlain));

        Assert.Equal(
            "无法连接",
            ColdStartReadiness.TechnicalDetail(cachePlain, "无法连接"));
    }
}
